### PR TITLE
Improving tests

### PR DESCRIPTION
**Pull request by diogobaeder**
Hi folks,

These changes first split unit and integration tests - we didn't have unit tests before, only integration -, and then I started adding some unit tests to increase the coverage of the production code. In this process, I also found a few issues, which I plan to fix later, after I can fully cover the library - I just didn't finish because I'm going to continue this next weekend.

One of the main reasons for covering the code with tests is because I want to refactor some parts of the code to improve it and make it more flexible and pythonic, and I can only confidently do that if I have the code covered.

I also introduced tests with Python 3.11 in the mix, to officially support it in the library - please make sure you have it available when running these tests.

### CODE REVIEW

File: .github/workflows/pinot_dbapi_test.yml
Comment (xiangfu0) [inline .github/workflows/pinot_dbapi_test.yml:40]: Shall we separate jobs from `pinot-dbapi-test` to `pinot-dbapi-unit-test` and `pinot-dbapi-integration-test`

**Review by xiangfu0 — APPROVED** [review]
lgtm! This is great!

File: .github/workflows/pinot_dbapi_test.yml
Comment (diogobaeder) [reply .github/workflows/pinot_dbapi_test.yml:40]: Yeah, good point, I'll do that now.

Comment (diogobaeder) [reply .github/workflows/pinot_dbapi_test.yml:40]: Done!

**Comment by diogobaeder** [comment]
Hi guys!

Anything else missing before these changes can be merged?

**Comment by walterddr** [comment]
> Hi guys!
> 
> Anything else missing before these changes can be merged?

thanks for the contribution @diogobaeder . could you give me some time to take a look at the PR? thanks.

File: tests/unit/test_db.py
Comment (walterddr) [inline tests/unit/test_db.py:139]: what exactly is this testing against? since the session is a MagicMock.

File: pinotdb/sqlalchemy.py
Comment (walterddr) [inline pinotdb/sqlalchemy.py:1]: great catch!

File: pinotdb/db.py
Comment (walterddr) [inline pinotdb/db.py:49]: could you be more specific? what's the goal of this TODO?

File: pinotdb/db.py
Comment (walterddr) [inline pinotdb/db.py:176]: yeah I think we can just delete this. no need for a todo right?

File: pinotdb/db.py
Comment (walterddr) [inline pinotdb/db.py:355]: can we just do the check here instead of adding a TODO.

**Review by walterddr — APPROVED** [review]
lgtm overall. plz kindly take a look at my comments and see if they make sense

**Comment by diogobaeder** [comment]
Sure thing, @walterddr ! Take as long as you need! I just asked because I saw the approval, so I thought it was good to be merged...

Let me address the specifics now. Thanks!

File: tests/unit/test_db.py
Comment (diogobaeder) [reply tests/unit/test_db.py:139]: This tests whether the Connection object is able to, generically, run a statement by delegating the request to the session object. Let me know if you'd prefer I add a docstring to explain the test a bit better.

File: pinotdb/sqlalchemy.py
Comment (diogobaeder) [reply pinotdb/sqlalchemy.py:1]: Thanks! :-)

File: pinotdb/db.py
Comment (diogobaeder) [reply pinotdb/db.py:49]: We should use functools.wraps here so that the decorated function doesn't lose its name. See here: https://docs.python.org/3/library/functools.html#functools.wraps - this is important to use when decorating a function, because, if not, then it might end up messing up with tools that depend on code inspection, like NewRelic agent etc.

File: pinotdb/db.py
Comment (diogobaeder) [reply pinotdb/db.py:176]: Sure, I just wanted to keep the mark there to reduce the scope of this particular PR. But if you want I can remove them right away.

File: pinotdb/db.py
Comment (diogobaeder) [reply pinotdb/db.py:355]: OK, I can add it to this PR. I wanted to avoid making the PR too big, but it's fine for me to fix that now.

**Comment by diogobaeder** [comment]
Thanks for the remarks, @walterddr ! I'll try to fix those points ASAP, maybe tomorrow. Overall, I left some TODOs there because I didn't want to overwhelm anyone with my changes, and keep the scope of these changes reduced to only code coverage as much as possible. The reason is, I'm not really very comfortable with fixing code that's not covered by tests, so I prefer to cover them first, then refactor/cleanup. My idea was to first try to hit 100% coverage, after a few PRs, and then start fixing or improving things. I can absolutely include some fixes with this PR though, no problem.

File: pinotdb/db.py
Comment (diogobaeder) [reply pinotdb/db.py]: Fixed!

File: pinotdb/db.py
Comment (diogobaeder) [reply pinotdb/db.py]: Fixed!

File: pinotdb/db.py
Comment (diogobaeder) [reply pinotdb/db.py]: Fixed!

File: tests/unit/test_db.py
Comment (diogobaeder) [reply tests/unit/test_db.py]: docstring added. Let me know what you think.

To be honest, I'm not completely satisfied with how the dependencies are played out there, so this test looks a bit weird - it's mocking something that is not a direct neighbor object to Connection. But I really prefer to first have the project covered with tests first, before I can take a stab at trying to improve the structure.

File: pinotdb/db.py
Comment (walterddr) [reply pinotdb/db.py]: Thanks for the clarification!

File: tests/unit/test_db.py
Comment (walterddr) [reply tests/unit/test_db.py]: Sounds good. Thanks. I saw similar test on other python dbapi projects as well just curious

**Comment by walterddr** [comment]
> Thanks for the remarks, @walterddr ! I'll try to fix those points ASAP, maybe tomorrow. Overall, I left some TODOs there because I didn't want to overwhelm anyone with my changes, and keep the scope of these changes reduced to only code coverage as much as possible. The reason is, I'm not really very comfortable with fixing code that's not covered by tests, so I prefer to cover them first, then refactor/cleanup. My idea was to first try to hit 100% coverage, after a few PRs, and then start fixing or improving things. I can absolutely include some fixes with this PR though, no problem.

Sounds good. Thank you for clarification. Yeah I agree with you splitting the PRs are usually more desired (also for rollbacks if any issue occurs). 

Forgive me if I were wrong on my remark on the complexity on some of the TODOs. I thought they are simpler thus I suggest addressing them in the same PR. I am ok with either a follow up or directly contributed in a single PR.

File: tests/unit/test_db.py
Comment (diogobaeder) [reply tests/unit/test_db.py]: Got it

File: pinotdb/db.py
Comment (walterddr) [inline pinotdb/db.py:51]: ahh, i see what you are saying. good catch

**Review by walterddr — APPROVED** [review]
lgtm

**Comment by walterddr** [comment]
lgtm could you resolve the conflict? we can merge afterwards

File: pinotdb/db.py
Comment (diogobaeder) [reply pinotdb/db.py:51]: :-)

**Comment by diogobaeder** [comment]
Conflicts solved, ready to merge! :-)

**Pull request merged**
